mongo/mongo.go

```diff
@@ -0,0 +1,362 @@
+	if err != nil {
+		return errors.Wrap(err, "error starting session")
+	}
+	defer session.EndSession(context.Background())
+
+	err = mongo.WithSession(context.TODO(), session, func(ctx mongo.SessionContext) error {
+		if err = session.StartTransaction(txnOptions); err != nil {
+			return errors.Wrap(err, "error: pending transaction")
+		}
+
+		err = db.executeTransactions(tx, ctx, session)
+		if err != nil {
+			return err
+		}
+
+		if err = errors.WithStack(session.CommitTransaction(context.Background())); err != nil {
+			return errors.Wrap(err, "error committing transaction")
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateTable creates a bucket or an embedded bucket if it does not exists.
+func (db *DB) createTable(bucket []byte, ctx context.Context) error {
+	if err := db.db.CreateCollection(ctx, string(bucket)); err != nil {
+		return errors.Wrap(err, "error creating collection")
+	}
+
+	// create an index on the Key field
+	index := mongo.IndexModel{
+		Keys:    createFilter("key", 1),
+		Options: options.Index().SetUnique(true),
+	}
+
+	_, err := db.db.Collection(string(bucket)).Indexes().CreateOne(context.TODO(), index)
+	if err != nil {
+		return errors.Wrap(err, "error creating collection")
+	}
+
+	return nil
+}
+
+// DeleteTable deletes a root or embedded bucket. Returns an error if the
+// bucket cannot be found or if the key represents a non-bucket value.
+func (db *DB) deleteTable(bucket []byte, ctx context.Context) error {
+	if !collectionExists(db.db, string(bucket)) {
+		return errors.Wrapf(database.ErrNotFound, "bucket %s not found", bucket)
+	}
+
+	if err := db.db.Collection(string(bucket)).Drop(ctx); err != nil {
+		return errors.Wrapf(err, "error dropping collection: %s", bucket)
+	}
+	return nil
+}
+
+func (db *DB) get(bucket, key []byte, ctx context.Context) (ret []byte, err error) {
+	filter := createFilter("key", key)
+	res := db.db.Collection(string(bucket)).FindOne(ctx, filter)
+
+	if err := res.Err(); err != nil {
+		return nil, errors.Wrapf(database.ErrNotFound, "%s/%s", bucket, key)
+	}
+
+	result := Tuple{}
+	if err := res.Decode(&result); err != nil {
+		return nil, errors.Wrap(err, "error decoding value")
+	}
+
+	return []byte(result.Value), nil
+}
+
+func (db *DB) set(bucket, key, value []byte, ctx context.Context) error {
+	filter := createFilter("key", key)
+	update := createUpdate(value)
+	opts := options.Update().SetUpsert(true)
+
+	_, err := db.db.Collection(string(bucket)).UpdateOne(ctx, filter, update, opts)
+	if err != nil {
+		return errors.Wrapf(err, "error setting value %s/%s", bucket, key)
+	}
+	return nil
+}
+
+// List returns the full list of entries in a bucket.
+func (db *DB) list(bucket []byte, ctx context.Context) ([]*database.Entry, error) {
+	if !collectionExists(db.db, string(bucket)) {
+		return nil, errors.Wrapf(database.ErrNotFound, "bucket %s not found", bucket)
+	}
+
+	// match all
+	filter := bson.D{{}}
```

Comment:
This can likely become `var filter bson.D`

Reply:
you mean just
```diff
+ var filter bson.D
- filter := bson.D{{}}
```

that throws the following error: `error 'failed listing values of collection testNoSQLUsers to: cannot marshal type primitive.D to a BSON Document: WriteNull can only write while positioned on a Element or Value but is positioned on a TopLevel' not expected`